### PR TITLE
remove pilot flag PILOT_TRACE_SAMPLING

### PR DESCRIPTION
**Pull request by zirain**
**Please provide a description of this PR:**

there's three ways to config trace sampling rate today, it's time to remove pilot flag.

### CODE REVIEW

File: pilot/pkg/networking/core/tracing.go [hunk 1]
```diff
@@ -577,8 +576,7 @@ func configureSampling(hcmTracing *hcm.HttpConnectionManager_Tracing, providerPe
 }
 
 func proxyConfigSamplingValue(config *meshconfig.ProxyConfig) float64 {
-	// PILOT_TRACE_SAMPLING
-	sampling := features.TraceSampling
+	sampling := 1.0
```
Comment (howardjohn) [inline pilot/pkg/networking/core/tracing.go:579]: how can this be turned off? `config.Tracing.Sampling != 0.0 {` does not allow it

**Review by howardjohn — COMMENTED** [review]
Given the widespread usage of this this may need a deprecation period to migrate users off of it. This is not just any feature flag, it is pretty extensively used -- and even has first class support in helm (`.Values.pilot.traceSampling`)

File: pilot/pkg/networking/core/tracing.go (hunk 1 above)
Comment (zirain) [reply pilot/pkg/networking/core/tracing.go:579]: what do you mean turned off? sampling = 0.0?

**Comment by zirain** [comment]
> Given the widespread usage of this this may need a deprecation period to migrate users off of it. This is not just any feature flag, it is pretty extensively used -- and even has first class support in helm (`.Values.pilot.traceSampling`)

what should I do for this? add a deprecation announcement in v1.23 releate notes?

**Review by hzxuzhonghu — COMMENTED** [review]
lGTM. maybe should deprecate sampling from ProxyConfig too.  ProxyConfig was designed not to work as an CR, since we have a replacement, it should end its life

**Comment by zirain** [comment]
> lGTM. maybe should deprecate sampling from ProxyConfig too. ProxyConfig was designed not to work as an CR, since we have a replacement, it should end its life

I don't think so, it work with defaultProviders as default value.

**Comment by hzxuzhonghu** [comment]
Telemetry cannot define default value?

**Comment by zirain** [comment]
> Telemetry cannot define default value?

https://github.com/istio/istio/pull/46711

File: releasenotes/notes/remove-pilot-tracing.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+apiVersion: release-notes/v2
+kind: feature
+area: telemetry
+releaseNotes:
+  - |
+    **Removed** pilot flag `PILOT_TRACE_SAMPLING`.
```
Comment (costinm) [inline releasenotes/notes/remove-pilot-tracing.yaml:6]: 'Instead users should use ????' ( I assume OTEL recommended env variables - but users shouldn't guess). A link to some doc (or wiki page or something ) would be nice too.

File: pilot/pkg/networking/core/tracing.go (hunk 1 above)
Comment (costinm) [reply pilot/pkg/networking/core/tracing.go:579]: The questions is what is the recommended way to configure trace sampling for pilot. I assumed using the OTel env - https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/ -  but this method is confusing me, shouldn't you remove this method and let OTel do its own config ?

File: releasenotes/notes/remove-pilot-tracing.yaml (hunk 2 above)
Comment (zirain) [reply releasenotes/notes/remove-pilot-tracing.yaml:6]: update upgrade notes

File: releasenotes/notes/remove-pilot-tracing.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+upgradeNotes:
+  - title: pilot flag `PILOT_TRACE_SAMPLING` supported in Istio.
```
Comment (hzxuzhonghu) [inline releasenotes/notes/remove-pilot-tracing.yaml:9]: removed

File: releasenotes/notes/remove-pilot-tracing.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+releaseNotes:
+  - |
+    **Removed** pilot flag `PILOT_TRACE_SAMPLING`.
+
+upgradeNotes:
+  - title: Removed pilot flag `PILOT_TRACE_SAMPLING` supported in Istio.
+    content: |
+      If you are using the pilot environment flag `PILOT_TRACE_SAMPLING`, you should
+      replace it with `defaultConfig.tracing.sampling` in the [MeshConfig](https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#ProxyConfig)
+      or `randomSamplingPercentage`in [Telemetry API](https://istio.io/latest/docs/reference/config/telemetry/).
```
Comment (costinm) [inline releasenotes/notes/remove-pilot-tracing.yaml:13]: - I don't see any code using Telemetry API to program Istiod itself. 
- TelemetryAPI is based on label selectors - Istiod can be external or run in another managed environment 
- The MeshConfig.tracing.sampling is supposed to be replaced by Telemetry API for sidecars - I don't think we should provide multiple options, we already have too many. If you want to use Telemetry API - we should just tell users to use the v1 API, not the alpha legacy one.
-  also the needs/expectations for Istiod are quite different from the 'default for all envoys in the mesh'

IMO we should just use the env variables defined by OTel - they are considered stable and that's how users expect to configure any app using the OTel SDK. 

Telemetry API would be the right thing to use for HTTP/gRPC straces - if Istiod was behind a Waypoint or Gateway. I am trying to bring back that mode ( I had a very old PR), Istiod is the perfect user of ambient and/or having a Gatweay in front, that would take care of this kind of tracing - but we still have the 'native' tracing using the SDK.

Comment (howardjohn) [reply releasenotes/notes/remove-pilot-tracing.yaml:13]: > I don't see any code using Telemetry API to program Istiod itself.

`PILOT_TRACE_SAMPLING` is not about programming Istiod-internal traces, its about setting the trace sampling for Envoys.

Comment (zirain) [reply releasenotes/notes/remove-pilot-tracing.yaml:13]: The name is misleading, I need to pause every time I read this code.

Comment (costinm) [reply releasenotes/notes/remove-pilot-tracing.yaml:13]: I don't know why I am even surprised by this after spending days trying to figure out the cert signing interfaces...

Can you put in some comment "for setting pilot tracing, we are using XXX ( or link to some doc - I remember John added OTLP tracing for pilot but forgot how to set it up...)" - and in the release note just mention the v1 telemetry API ( not mesh config )

Comment (zirain) [reply releasenotes/notes/remove-pilot-tracing.yaml:13]: Let me try to find out thing about OTLP tracing for pilot(I never know about it).

meshconfig is still useful when users use defaultProviers only(without Telemetry resource).

**Pull request closed without merging**